From 9c665ae3c4684f8826f4b7d8e0a75b9a986e8617 Mon Sep 17 00:00:00 2001
From: Gunter Labes <soliton@wesnoth.org>
Date: Sun, 29 Sep 2024 16:45:02 +0200
Subject: [PATCH] Use completion token overload to fix compile issues with
 boost 1.86

Keep behavior of previous (now deprecated) overload by rethrowing exceptions.

Fixes #9284.
---
 src/server/common/server_base.cpp | 21 ++++++++++++++++++---
 src/server/wesnothd/server.cpp    | 15 +++++++++++++--
 2 files changed, 31 insertions(+), 5 deletions(-)

diff --git a/src/server/common/server_base.cpp b/src/server/common/server_base.cpp
index 1ed63200d7ad..222cc086cfe1 100644
--- a/src/server/common/server_base.cpp
+++ b/src/server/common/server_base.cpp
@@ -74,10 +74,18 @@ server_base::server_base(unsigned short port, bool keep_alive)
 void server_base::start_server()
 {
 	boost::asio::ip::tcp::endpoint endpoint_v6(boost::asio::ip::tcp::v6(), port_);
-	boost::asio::spawn(io_service_, [this, endpoint_v6](boost::asio::yield_context yield) { serve(yield, acceptor_v6_, endpoint_v6); });
+	boost::asio::spawn(io_service_, [this, endpoint_v6](boost::asio::yield_context yield) { serve(yield, acceptor_v6_, endpoint_v6); }
+#if BOOST_VERSION >= 108000
+		, [](std::exception_ptr e) { if (e) std::rethrow_exception(e); }
+#endif
+	);
 
 	boost::asio::ip::tcp::endpoint endpoint_v4(boost::asio::ip::tcp::v4(), port_);
-	boost::asio::spawn(io_service_, [this, endpoint_v4](boost::asio::yield_context yield) { serve(yield, acceptor_v4_, endpoint_v4); });
+	boost::asio::spawn(io_service_, [this, endpoint_v4](boost::asio::yield_context yield) { serve(yield, acceptor_v4_, endpoint_v4); }
+#if BOOST_VERSION >= 108000
+		, [](std::exception_ptr e) { if (e) std::rethrow_exception(e); }
+#endif
+	);
 
 	handshake_response_ = htonl(42);
 
@@ -115,7 +123,11 @@ void server_base::serve(boost::asio::yield_context yield, boost::asio::ip::tcp::
 	}
 
 	if(accepting_connections()) {
-		boost::asio::spawn(io_service_, [this, &acceptor, endpoint](boost::asio::yield_context yield) { serve(yield, acceptor, endpoint); });
+		boost::asio::spawn(io_service_, [this, &acceptor, endpoint](boost::asio::yield_context yield) { serve(yield, acceptor, endpoint); }
+#if BOOST_VERSION >= 108000
+			, [](std::exception_ptr e) { if (e) std::rethrow_exception(e); }
+#endif
+		);
 	} else {
 		return;
 	}
@@ -560,6 +572,9 @@ template<class SocketPtr> void server_base::async_send_doc_queued(SocketPtr sock
 		io_service_, [this, doc_ptr = doc.clone(), socket](boost::asio::yield_context yield) mutable {
 			send_doc_queued(socket, doc_ptr, yield);
 		}
+#if BOOST_VERSION >= 108000
+		, [](std::exception_ptr e) { if (e) std::rethrow_exception(e); }
+#endif
 	);
 }
 
diff --git a/src/server/wesnothd/server.cpp b/src/server/wesnothd/server.cpp
index a813900ba9c4..c208a8aed22f 100644
--- a/src/server/wesnothd/server.cpp
+++ b/src/server/wesnothd/server.cpp
@@ -666,12 +666,20 @@ void server::refresh_tournaments(const boost::system::error_code& ec)
 
 void server::handle_new_client(socket_ptr socket)
 {
-	boost::asio::spawn(io_service_, [socket, this](boost::asio::yield_context yield) { login_client(yield, socket); });
+	boost::asio::spawn(io_service_, [socket, this](boost::asio::yield_context yield) { login_client(yield, socket); }
+#if BOOST_VERSION >= 108000
+		, [](std::exception_ptr e) { if (e) std::rethrow_exception(e); }
+#endif
+	);
 }
 
 void server::handle_new_client(tls_socket_ptr socket)
 {
-	boost::asio::spawn(io_service_, [socket, this](boost::asio::yield_context yield) { login_client(yield, socket); });
+	boost::asio::spawn(io_service_, [socket, this](boost::asio::yield_context yield) { login_client(yield, socket); }
+#if BOOST_VERSION >= 108000
+		, [](std::exception_ptr e) { if (e) std::rethrow_exception(e); }
+#endif
+	);
 }
 
 template<class SocketPtr>
@@ -777,6 +785,9 @@ void server::login_client(boost::asio::yield_context yield, SocketPtr socket)
 
 	boost::asio::spawn(io_service_,
 		[this, socket, new_player](boost::asio::yield_context yield) { handle_player(yield, socket, new_player); }
+#if BOOST_VERSION >= 108000
+		, [](std::exception_ptr e) { if (e) std::rethrow_exception(e); }
+#endif
 	);
 
 	LOG_SERVER << log_address(socket) << "\t" << username << "\thas logged on"
